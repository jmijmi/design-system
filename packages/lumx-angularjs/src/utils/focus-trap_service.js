import { TAB_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function FocusTrapService() {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The active element where to trap the focus.
     *
     * @type {Element}
     */
    let _activeElement;

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle key events on key down.
     *
     * @param {Event} evt The key event.
     */
    function _onKeyDown(evt) {
        if (evt.keyCode !== TAB_KEY_CODE) {
            return;
        }

        const focusableEls = _activeElement.find(
            'a[href]:not([tabindex="-1"]), button:not([tabindex="-1"]), textarea:not([tabindex="-1"]), input[type="text"]:not([tabindex="-1"]), input[type="radio"]:not([tabindex="-1"]), input[type="checkbox"]:not([tabindex="-1"]), [tabindex]:not([tabindex="-1"])',
        );

        const firstFocusableEl = focusableEls[0];
        const lastFocusableEl = focusableEls[focusableEls.length - 1];

        if (evt.shiftKey) {
            if (document.activeElement === firstFocusableEl) {
                lastFocusableEl.focus();
                evt.preventDefault();
            }
        } else if (document.activeElement === lastFocusableEl) {
            firstFocusableEl.focus();
            evt.preventDefault();
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Activate focus trap on given element.
     *
     * @param {element} el The element where to trap focus.
     */
    function activate(el) {
        _activeElement = el;
        _activeElement.on('keydown', _onKeyDown);
    }

    /**
     * Disable focus trap on given element.
     */
    function disable() {
        _activeElement.off('keydown', _onKeyDown);
        _activeElement = undefined;
    }

    /////////////////////////////

    service.activate = activate;
    service.disable = disable;
}

/////////////////////////////

angular.module('lumx.utils.focus-trap').service('LxFocusTrapService', FocusTrapService);

/////////////////////////////

export { FocusTrapService };
