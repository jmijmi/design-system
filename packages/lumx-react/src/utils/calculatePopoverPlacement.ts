import { Placement } from '@lumx/react/components/popover/Popover';

export const calculatePopoverPlacement = (
    placement: Placement,
    anchor: ClientRect,
    popover: ClientRect,
): {
    x: number;
    y: number;
} => {
    const splittedPlacement = placement.split('-');
    const verticalPlacement = splittedPlacement[0];
    const horizontalPlacement = splittedPlacement[1];
    const { top, left, width: widthAnchor, height: heightAnchor } = anchor;
    const { width: widthPopover, height: heightPopover }: ClientRect | DOMRect = popover;
    let x = 0;
    let y = 0;
    switch (verticalPlacement) {
        case 'top':
            y += top - heightPopover;

            switch (horizontalPlacement) {
                case 'start':
                    x += left;

                    break;
                case 'end':
                    x += left + widthAnchor - widthPopover;
                    break;

                default:
                    x += left + (widthAnchor - widthPopover) / 2;

                    break;
            }

            break;
        case 'right':
            x += left + widthAnchor;

            switch (horizontalPlacement) {
                case 'start':
                    y += top;

                    break;
                case 'end':
                    y += top + heightAnchor - heightPopover;

                    break;

                default:
                    y += top + (heightAnchor - heightPopover) / 2;

                    break;
            }

            break;
        case 'bottom':
            y += top + heightAnchor;

            switch (horizontalPlacement) {
                case 'start':
                    x += left;

                    break;
                case 'end':
                    x += left + widthAnchor - widthPopover;

                    break;

                default:
                    x += left + (widthAnchor - widthPopover) / 2;

                    break;
            }

            break;
        case 'left':
            x += left - widthPopover;

            switch (horizontalPlacement) {
                case 'start':
                    y += top;

                    break;
                case 'end':
                    y += top + heightAnchor - heightPopover;

                    break;

                default:
                    y += top + (heightAnchor - heightPopover) / 2;

                    break;
            }

            break;

        default:
            break;
    }

    return { x, y };
};
