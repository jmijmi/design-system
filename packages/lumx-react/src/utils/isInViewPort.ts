export type ViewPortVisibility = 'full' | 'partial';

/**
 * Test if an element is in the viewport.
 *
 * @param bounding Bounding of element to test.
 * @param visibility Visible only partially or fully.
 * @return Whether the element to test is in viewport.
 */
export const isInViewPort = (bounding: ClientRect | DOMRect, visibility: ViewPortVisibility): boolean => {
    if (visibility === 'partial') {
        return (
            bounding.top <= (window.innerHeight || document.documentElement.clientHeight) &&
            bounding.left <= (window.innerWidth || document.documentElement.clientWidth) &&
            bounding.right >= 0 &&
            bounding.bottom >= 0
        );
    }

    return (
        bounding.top >= 0 &&
        bounding.left >= 0 &&
        bounding.right <= (window.innerWidth || document.documentElement.clientWidth) &&
        bounding.bottom <= (window.innerHeight || document.documentElement.clientHeight)
    );
};
