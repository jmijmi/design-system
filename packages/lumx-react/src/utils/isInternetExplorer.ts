/**
 * Check if browser is IE
 * @return Browser is IE or not
 */
const isInternetExplorer = () => {
    const ua = window.navigator.userAgent;
    const msie = ua.indexOf('MSIE ');
    const isIEVersion = !!navigator.userAgent.match(/Trident.*rv\:11\./);
    return msie > 0 || isIEVersion;
};

export { isInternetExplorer };
