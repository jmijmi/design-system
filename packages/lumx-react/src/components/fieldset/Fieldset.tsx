import React, { ReactElement } from 'react';

import classNames from 'classnames';

import { COMPONENT_PREFIX } from '@lumx/react/constants';
import { IGenericProps, getRootClassName } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
interface IFieldsetProps extends IGenericProps {
    legend?: ReactElement | string;
}

const COMPONENT_NAME = `${COMPONENT_PREFIX}Fieldset`;

/**
 * The default class name and classes prefix for this component.
 *
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

const Fieldset: React.FC<IFieldsetProps> = ({
    className = '',
    children,
    legend,
    ...props
}: IFieldsetProps): React.ReactElement => (
    <fieldset className={classNames(className, CLASSNAME)} {...props}>
        {legend && <legend className={`${CLASSNAME}__legend`}>{legend}</legend>}
        {children}
    </fieldset>
);
Fieldset.displayName = COMPONENT_NAME;

export { CLASSNAME, Fieldset, IFieldsetProps };
