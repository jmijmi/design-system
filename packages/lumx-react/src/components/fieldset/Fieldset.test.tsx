import { ShallowWrapper, shallow } from 'enzyme';
import 'jest-enzyme';
import React from 'react';

import { ICommonSetup } from '@lumx/react/testing/utils';
import { CLASSNAME, Fieldset, IFieldsetProps } from './Fieldset';

interface ISetup extends ICommonSetup {
    children: ShallowWrapper;
    fieldset: ShallowWrapper;
    legend: ShallowWrapper;
}

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param propOverrides An object that will extend the default properties.
 * @return An object with some shortcuts to elements or data required in tests.
 */
const setup = (propOverrides: Partial<IFieldsetProps> = {}): ISetup => {
    const props = {
        children: <input type="text" name="input" />,
        legend: 'I am a legend',
        ...propOverrides,
    };
    const wrapper = shallow(<Fieldset {...props} />);
    const fieldset = wrapper.find('fieldset').first();

    return {
        children: fieldset.find('input'),
        fieldset,
        legend: fieldset.find('legend').first(),
        props,
        wrapper,
    };
};

describe('<Fieldset />', () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshot', () => {
        it('should render correctly Chip component', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { fieldset, legend, children }: ISetup = setup();

            expect(fieldset).toHaveClassName(CLASSNAME);
            expect(legend).toHaveLength(1);
            expect(children).toHaveLength(1);
        });

        it('should use the given text as legend', (): void => {
            const legendValue = 'I am a legend';
            const { legend }: ISetup = setup({
                legend: legendValue,
            });

            expect(legend).toHaveText(legendValue);
        });

        it('should use the given html as legend', (): void => {
            const legendContent = <span className="lumapps">Be bold</span>;
            const { legend }: ISetup = setup({
                legend: legendContent,
            });

            expect(legend).toHaveLength(1);
            expect(legend).toContainReact(legendContent);
        });

        it('should not have any child', (): void => {
            const { children }: ISetup = setup({
                children: null,
            });

            expect(children).toHaveLength(0);
        });

        it('should use the given className prop', (): void => {
            const { fieldset }: ISetup = setup({
                className: 'my-fieldset',
            });

            expect(fieldset).toHaveClassName(CLASSNAME);
            expect(fieldset).toHaveClassName('my-fieldset');
        });
    });

    // 3. Test events.
    // NA

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        it('should not render the legend if there is no given', () => {
            const { legend } = setup({ legend: '' });
            expect(legend).toHaveLength(0);
        });
    });

    // 5. Test state.
    // N/A.
});
