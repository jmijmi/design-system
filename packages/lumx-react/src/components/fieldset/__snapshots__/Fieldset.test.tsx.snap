// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fieldset /> Snapshot should render correctly Chip component 1`] = `
<fieldset
  className="lumx-fieldset"
>
  <legend
    className="lumx-fieldset__legend"
  >
    I am a legend
  </legend>
  <input
    name="input"
    type="text"
  />
</fieldset>
`;
